### DISCUSSION TITLE
Add Python version datasource

### DISCUSSION THREAD

**Discussion by not7cd** [renovatebot/renovate#27581] [Suggest an Idea]
### Tell us more.

Just like for [Node releases](https://docs.renovatebot.com/node/), Renovate should be able to propose update to Python version. I want Renovate to bump deprecated Python versions that are hardcoded as variable in Dockerfiles. This can be done with custom managers, but a release data source is needed first. 

Python.org offers an API with all releases: https://www.python.org/api/v2/downloads/release/

While release status is found here: https://github.com/python/devguide/blob/main/include/release-cycle.json

**Comment by jamietanna** [comment]
We should hopefully be able to wire https://docs.renovatebot.com/modules/datasource/endoflife-date/ in with this to source deprecation / EOL dates

**Comment by rarkins** [comment]
It's important that the source chosen is also representative of whether it can be installed. In this case it's essential that it does not "precede" Renovate's ability to install python using `install-tool`, otherwise you'll get a PR updating python with an artifacts error because the new Python error wasn't installable. That source is here: https://github.com/containerbase/base/blob/main/src/usr/local/containerbase/utils/python.sh

@viceice is it using virtualenv to install python versions?

**Reply by rarkins** [reply]
We use the following datasource when it comes to calling install-tool within Renovate: https://github.com/renovatebot/renovate/blob/3f315fb88159780f537e3b24ba1bebcebab406c4/lib/util/exec/containerbase.ts#L164-L168

**Reply by rarkins** [reply]
If your new datasource is "ahead" of containerbase/python-prebuild then artifact updates will fail because install-tool python will fail

**Reply by not7cd** [reply]
Do you suggest clamping releases to what is available in `containerbase/python-prebuild`?

**Reply by not7cd** [reply]
I see that everything except pre releases is present in https://github.com/containerbase/python-prebuild/releases

**Reply by not7cd** [reply]
Comparing https://github.com/containerbase/python-prebuild/releases/tag/3.12.2 and https://www.python.org/downloads/release/python-3122/ I can see a two-day difference between releases. So this is an edge case that I imagine this Datasource must handle?

**Reply by rarkins** [reply]
It's not an easy decision:
- You can take a "real time" datasource but then face the problem that you update projects into versions they can't install yet, or
- You take a slower datasource to ensure it can be installed, but then users may wonder why the new release isn't proposed yet

The python-prebuild releases seem to come from pyenv: https://github.com/containerbase/python-prebuild/blob/3.12.2/bin/install-builder.sh

So then it seems the path is python release => pyenv support => containerbase support

**Reply by not7cd** [reply]
pyenv support is updated manually. https://github.com/pyenv/pyenv/pull/2899
This may be interpreted as an index
https://github.com/pyenv/pyenv/tree/master/plugins/python-build/share/python-build

**Reply by not7cd** [reply]
This has already bitten someone in https://github.com/kachick/renovate-config-asdf/pull/1568 using [asdf manager](https://docs.renovatebot.com/modules/manager/asdf/)

**Reply by viceice** [reply]
I've already a hack to try to build a new python patch from previous version when pyenv isn't updated.

but that doesn't work for minors

**Reply by rarkins** [reply]
Wondering:
- Is there an alternative to pyenv to allow faster builds of new releases?
- Is pyenv holding back the prereleases or is that our own policy in containerbase?

**Reply by not7cd** [reply]
I would opt for slower _containerbase/python-prebuild_ dependent release. Because this is the final bottleneck. This Datasource will benefit from any optimizations to it.

**Comment by not7cd** [comment]
I have created a PR that attempts to implement this data source based on `node-version`
https://github.com/renovatebot/renovate/pull/27583

**Reply by not7cd** [reply]
This PR is ready. To no surprise, I'm now in the process of getting rid of deprecated Python versions. And I'm going to test it in "production" in the next week.

**Comment by shadycuz** [comment]
I had a similar discussion https://github.com/renovatebot/renovate/discussions/22222

But this was about in src repositories, not docker files. 

I actually still need this feature =/, but maybe I should contribute instead of complain https://github.com/renovatebot/renovate/issues/22225
